R/ChromBackendDataFrame-functions.R

```diff
@@ -0,0 +1,152 @@
+                          paste(missing_cn, collapse = ", "),
+                          " is/are missing"))
+    }
+    NULL
+}
+
+#' Function to check data types of selected columns in the provided `DataFrame`.
+#'
+#' @param x `DataFrame` to validate.
+#'
+#' @param datatypes named `character`, names being column names and elements
+#'     expected data types.
+#'
+#' @author Johannes Rainer
+#'
+#' @noRd
+.valid_column_datatype <- function(x, datatypes = .CHROMATOGRAMS_DATA_COLUMNS) {
+    datatypes <- datatypes[names(datatypes) %in% colnames(x)]
+    res <- mapply(FUN = .is_class, x[, names(datatypes), drop = FALSE],
+                  datatypes)
+    if (!all(res))
+        paste0("The following columns have a wrong data type: ",
+               paste(names(res[!res]), collapse = ", "),
+               ". The expected data type(s) is/are: ",
+               paste(datatypes[names(res)[!res]], collapse = ", "), ".")
+    else NULL
+}
+
+#' @importFrom MsCoreUtils vapply1l
+#'
+#' @noRd
+.valid_rtime_column <- function(x) {
+    if (length(x$rtime)) {
+        if (!all(vapply1l(x$rtime, is.numeric)))
+            return("rtime column should contain a list of numeric")
+        if (any(vapply1l(x$rtime, is.unsorted)))
+            return("rtime values have to be sorted increasingly")
+    }
+    NULL
+}
+
+.valid_intensity_column <- function(x) {
+    if (length(x$intensity))
+        if (!all(vapply1l(x$intensity, is.numeric)))
+            return("intensity column should contain a list of numeric")
+    NULL
+}
+
+.valid_intensity_rtime_columns <- function(x) {
+    ## Don't want to have that tested on all on-disk objects.
+    if (length(x$intensity) && length(x$rtime))
+        if (any(lengths(x$intensity) != lengths(x$rtime)))
+            return(paste0("Length of rtime and intensity values differ for",
+                          " some chromatograms"))
+    NULL
+}
+
+.CHROMATOGRAMS_DATA_COLUMNS <- c(
+    dataOrigin = "character",
+    dataStorage = "character",
+    intensity = "NumericList",
+    msLevel = "integer",
+    mz = "numeric",
+    mzMin = "numeric",
+    mzMax = "numeric",
+    precursorMz = "numeric",
+    precursorMzMin = "numeric",
+    precursorMzMax = "numeric",
+    productMz = "numeric",
+    productMzMin = "numeric",
+    productMzMax = "numeric",
+    rtime = "NumericList"
+)
+
+#' @rdname ChromBackend
+#'
+#' @export ChromBackendDataFrame
+ChromBackendDataFrame <- function() {
+    new("ChromBackendDataFrame")
+}
+
+#' Helper function to return a column from the (chrom data) `DataFrame`. If
+#' the column `column` is an `Rle` `as.vector` is called on it. If column is
+#' the name of a mandatory variable but it is not available it is created on
+#' the fly.
+#'
+#' @param x `DataFrame`
+#'
+#' @param column `character(1)` with the name of the column to return.
+#'
+#' @importMethodsFrom S4Vectors [[
+#'
+#' @author Johannes Rainer
+#'
+#' @noRd
+.get_rle_column <- function(x, column) {
```

Comment:
I had a similar thought. But then I decided to not move this function because of the `.CHROMATOGRAMS...`. I have no strong opinion on that, but preferentially I'd like to keep it as it is.

Reply:
Ok, I just want to avoid to repeat ourselves (and to create multiple locations for bug fixing; maybe we should add a comment: similar code in `Spectra`. Just in case we find a bug and forgot that the code was duplicated).

Reply:
Sure. I'll add a note here and in `Spectra`.

Reply:
Done.